Tasks/NpmAuthenticate/npmauth.ts

```diff
@@ -0,0 +1,136 @@
+import * as tl from 'vsts-task-lib/task';
+import * as URL from 'url';
+import * as fs from 'fs';
+import * as constants_1 from './constants';
+import * as npmregistry_1 from 'npm-common/npmregistry'
+import * as util from 'npm-common/util';
+
+async function main(): Promise<void> {
+    tl.setResourcePath(path.join(__dirname, 'task.json'));
+    let saveNpmrcPath;
+    let npmrc = tl.getInput(constants_1.NpmAuthenticateTaskInput.WorkingDir);
+    let workingDir = path.dirname(npmrc);
+    if (!(npmrc.endsWith('.npmrc') && tl.exist(npmrc))) {
+        throw new Error(tl.loc('NpmrcNotNpmrc', npmrc));
+    }
+    else {
+        console.log(tl.loc("AuthenticatingThisNpmrc", npmrc));
+    }
+
+    if (process.env.SAVE_NPMRC_PATH) {
+         saveNpmrcPath = process.env.SAVE_NPMRC_PATH
+    }
+    else {
+        let tempPath = tl.getVariable('Agent.BuildDirectory') || tl.getVariable('Agent.ReleaseDirectory') || process.cwd();
+        tempPath = path.join(tempPath, 'npmAuthenticate');
+        tl.mkdirP(tempPath);
+        saveNpmrcPath = fs.mkdtempSync(tempPath + path.sep); 
+        console.log("##vso[task.setvariable variable=SAVE_NPMRC_PATH;]" + saveNpmrcPath);
+        console.log("##vso[task.setvariable variable=NPM_AUTHENTICATE_TEMP_DIRECTORY;]" + tempPath);
+        
+    }
+    let npmrcTable;
+
+    //The index file is a json object that keeps track of where .npmrc files are saved.
+    //There is a key-value pairing of filepaths of original npmrc files to IDs.
+    //This is important so multiple runs of the npm Authenticate task on the same .npmrc file actually reverts to the original after the build completes.
+    let indexFile = saveNpmrcPath+'\\index.json';
+
+    if (fs.existsSync(indexFile)) { //If the file exists, add to it.
+        npmrcTable = JSON.parse(fs.readFileSync(indexFile, 'utf8'));
+        if (npmrcTable[npmrc]=== undefined) {
+            npmrcTable[npmrc] = npmrcTable['index'];
+            npmrcTable['index']++;
+            fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
+            util.saveFileWithName(npmrc, npmrcTable[npmrc], saveNpmrcPath);
+        }
+    }
+    else { //If the file doesn't exist, create it. 
+        npmrcTable = new Object();
+        npmrcTable['index']=0;
+        npmrcTable[npmrc] = npmrcTable['index'];
+        npmrcTable['index']++;
+        fs.writeFileSync(indexFile, JSON.stringify(npmrcTable));
+        util.saveFileWithName(npmrc, '0', saveNpmrcPath);
+
+    }
+
+    let endpointRegistries;
+    let endpointIds = tl.getDelimitedInput(constants_1.NpmAuthenticateTaskInput.CustomEndpoint, ',');
+    if (endpointIds && endpointIds.length > 0) {
+        endpointRegistries = endpointIds.map(e => npmregistry_1.NpmRegistry.FromServiceEndpoint(e, true));
+    }
+    let LocalNpmRegistries = await util.getLocalNpmRegistries(workingDir);
+    
+    
+    let npmrcFile = fs.readFileSync(npmrc, 'utf8').split('\n');
+    for (let fileLine of npmrcFile) {
```

Comment:
`GetRegistries` on `npmrcparser.ts` should give you a list of all registries in an .npmrc file. It should be what you need so no need to re-implement this.